### DISCUSSION TITLE
Issues with Next.js + Prisma + GraphQL

### DISCUSSION THREAD

**Discussion by cbaucom** [MichalLytek/typegraphql-prisma#122] [Q&A]
Hi, I'm having trouble getting my app stable using Next, Prisma, GraphQL, and TypeGraphQL. There are 2 issues that I can't seem to figure out and they seem to be related to Babel. 

Here is my `.babelrc`: 
```
{
  "presets": ["@babel/preset-typescript", "next/babel"],
  "plugins": [
    "babel-plugin-transform-typescript-metadata",
    "babel-plugin-transform-es2015-modules-commonjs",
    ["@babel/plugin-proposal-decorators", { "legacy": true }],
    ["@babel/plugin-proposal-class-properties", { "loose": true }]
  ]
}
```

Without `"babel-plugin-transform-es2015-modules-commonjs",` I get a 500 error at my graphql endpoint. 
With `"babel-plugin-transform-es2015-modules-commonjs",` I get `Uncaught ReferenceError: exports is not defined`. 

Here is a [sample repo](https://github.com/cbaucom/nextjs-apollo-typegraphql-prisma-demo) to check out and reproduce the errors. 

Here is the list of my dependencies for quick review: 
```
"dependencies": {
    "@apollo/client": "^3.1.1",
    "@babel/plugin-proposal-class-properties": "^7.10.4",
    "@babel/plugin-proposal-decorators": "^7.10.5",
    "@prisma/client": "~2.23.0",
    "@types/graphql-fields": "^1.3.3",
    "apollo-server-micro": "^2.16.1",
    "babel-plugin-transform-typescript-metadata": "^0.3.0",
    "class-validator": "^0.13.1",
    "graphql": "^15.5.0",
    "graphql-fields": "^2.0.3",
    "graphql-scalars": "^1.9.3",
    "graphql-type-json": "^0.3.2",
    "next": "^10.2.3",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "reflect-metadata": "^0.1.13",
    "type-graphql": "^1.1.1"
  },
  "devDependencies": {
    "@babel/core": "^7.14.3",
    "@babel/preset-typescript": "^7.13.0",
    "@types/faker": "^5.5.5",
    "@types/react": "^17.0.8",
    "babel-plugin-transform-es2015-modules-commonjs": "^6.26.2",
    "cross-env": "^7.0.2",
    "faker": "^5.5.3",
    "prisma": "^2.23.0",
    "ts-node": "^10.0.0",
    "typegraphql-prisma": "^0.14.3",
    "typescript": "^4.3.2"
  }
```

If I should post any more info here please let me know, it should all be in the repo though. 

Any help would be much appreciated.

**Comment by MichalLytek** [comment]
Please try the tips from this thread:

https://github.com/MichalLytek/type-graphql/issues/55#issuecomment-749603595

**Reply by cbaucom** [reply]
I just tried with the babel suggestion there but no luck. Still getting a 500 error for the graphql endpoint and this error in the dev console locally 
```
ReferenceError: Cannot access 'ItemWhereInput' before initialization
    at Module.ItemWhereInput (webpack-internal:///./graphql/generated/type-graphql/resolvers/inputs/ItemWhereInput.ts:3:55)
```

**Reply by MichalLytek** [reply]
Ah, so unfortunately there's not fix for that for now 😞 See #2

**Reply by cbaucom** [reply]
Any ideas how I can make typegraphql-prisma + nextjs + graphql work together? I tried this below but no dice. 

> I've been a full day trying to solve this problem and finally, I had to use "babel-plugin-transform-es2015-modules-commonjs" babel plugin.

When I try adding this to my project, I then get a 500 error for my graphql endpoint instead of the original error this thread is about. Any ideas how to then fix the 500 error for /api/graphql? GH repo [here](https://github.com/cbaucom/nextjs-apollo-typegraphql-prisma-demo)

**Reply by MichalLytek** [reply]
Maybe try to check the error message as `500` is not enough info to help you.

**Reply by cbaucom** [reply]
Yeah, it just goes back to that ReferenceError that is still an open issue. I was trying to follow along with the 2 people that seem to have got something working and one was adding `babel-plugin-transform-es2015-modules-commonjs` to `.babelrc` but that just causes an `exports not defined` issue. 

The other suggestion from that thread said "transpile the es6 to es5 with babel after compiling typescript. Not sure if this is a good practice but seems to work now." - either I don't know how to do this properly or it doesn't work either. 

Any chance you have time to play around with the sample repo above to see if you can get something working locally?

**Reply by MichalLytek** [reply]
Nope, I have no experience with Next.js so I can't help you with that.

**Comment by MichalLytek** [comment]
What about existing example repos?
https://github.com/jaysc/NextJS-Apollo-TypeGraphQL-Prisma

**Reply by cbaucom** [reply]
That's the exact repo I started with. I can look again...

**Comment by taylorpreston** [comment]
Hey! Has anyone had success on getting this working yet? I recently learned about TypeGraphQL and am really excited about how easy it could make some of my side projects 😄

**Comment by SheaBelsky** [comment]
I got this to work a while ago, and put a boilerplate for it into a GitHub repo: 

https://github.com/SheaBelsky/nextjs-prisma-typegraphql

It's been a while so I can't promise it works 100%, but things were operational last I checked. The only thing I think I was working on before was working on bundle analysis, I think it was bigger than normal and causing some client-side issues.
